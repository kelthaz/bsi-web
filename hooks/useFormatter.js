import phoneFormatter from '../helpers/phoneFormatter';
import uppercaseFormatter from '../helpers/uppercaseFormatter';

const useFormatter = (format) => {
  switch (format) {
    case 'phone':
      return { formatter: phoneFormatter, changeSelection: true };

    case 'uppercase':
      return { formatter: uppercaseFormatter, changeSelection: true };

    default:
      return (value) => value;
  }
};

export default useFormatter;
