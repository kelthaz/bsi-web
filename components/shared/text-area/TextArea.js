import React from 'react';
import PropTypes from 'prop-types';
import styles from './text-area.module.scss';
import useFormatter from '../../../hooks/useFormatter';
import { regexMultipleSpaces } from '../../../constants/regex';

const TextArea = ({ name, error, touched, value, setValue, setTouched, label, optional, maxlength, format }) => {
  const [formatter, changeSelection] = useFormatter(format);

  const onBeforeInput = (event) => {
    if (formatter(event.data) === '') {
      event.preventDefault();
    }
  };

  const onTouch = async () => {
    if (!touched) {
      await setTouched(true);
    }
  };

  const onChange = async (event) => {
    const { selectionStart, value: valueTarget } = event.target;
    const formattedValue = formatter(valueTarget.trimStart().replace(regexMultipleSpaces, ' '));
    await onTouch();
    await setValue(formattedValue);
    changeSelection('', '', valueTarget, formattedValue, selectionStart, event.target);
  };

  const onBlur = async () => {
    await setValue(value.trimEnd());
    await onTouch();
  };

  const hasError = touched && error;

  return (
    <div className={styles['text-area-container']}>
      <div className={`${styles['border-text-area']} ${hasError && styles['indicador-error']}`}>
        <textarea
          id={name}
          name={name}
          className={styles['text-area']}
          onChange={onChange}
          onBlur={onBlur}
          value={value}
          maxLength={maxlength}
          autoComplete="off"
          placeholder={label}
          tabIndex="0"
          onBeforeInput={onBeforeInput}
        />

        <div className={styles['resize-text-area']} />
      </div>

      <span className={hasError ? styles['help-text-error'] : styles['help-text']}>
        {hasError ? error : optional && 'Opcional'}&nbsp;
      </span>
    </div>
  );
};

TextArea.propTypes = {
  label: PropTypes.any.isRequired,
  name: PropTypes.string.isRequired,
  format: PropTypes.string.isRequired,
  error: PropTypes.string,
  touched: PropTypes.any.isRequired,
  value: PropTypes.any.isRequired,
  setValue: PropTypes.any.isRequired,
  setTouched: PropTypes.any.isRequired,
  maxlength: PropTypes.number.isRequired,
  optional: PropTypes.bool,
};

TextArea.defaultProps = {
  optional: false,
  error: '',
};

export default TextArea;
