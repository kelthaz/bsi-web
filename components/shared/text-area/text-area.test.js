import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import useFormatter from '../../../hooks/useFormatter';
import TextArea from './TextArea';

jest.mock('../../../hooks/useFormatter', () => jest.fn());

describe('Pruebas en el componente TextArea', () => {
  const setValue = jest.fn();
  const setTouched = jest.fn();
  const changeSelection = jest.fn();

  beforeEach(() => {
    useFormatter.mockReturnValue([jest.fn((value) => value), changeSelection]);
    jest.clearAllMocks();
  });

  test('Debe de mostrarse correctamente el componente', () => {
    // arrange
    const props = {
      name: 'tellUs',
      touched: false,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    // act
    const wrapper = shallow(<TextArea {...props} />);
    // assert
    expect(wrapper).toMatchSnapshot();
  });

  it('Debe de llamarse el setValue y el setTouched, con el valor nuevo', async () => {
    // arrange
    const valorEsperado = 'soy una empresa que vende empanadas';
    const props = {
      name: 'tellUs',
      touched: false,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    await act(async () => {
      wrapper
        .find('textarea')
        .simulate('change', { target: { value: valorEsperado, selectionStart: 0, setSelectionRange: jest.fn() } });
    });

    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado);
    expect(setTouched).toHaveBeenCalledTimes(1);
    expect(setTouched).toHaveBeenCalledWith(true);
  });

  test('Debe de llamarse el setValue y no el setTouched, con el valor nuevo', async () => {
    // arrange
    const valorEsperado = 'soy una empresa que vende empanadas en cali';
    const props = {
      name: 'tellUs',
      touched: true,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    await act(async () => {
      wrapper
        .find('textarea')
        .simulate('change', { target: { value: valorEsperado, selectionStart: 0, setSelectionRange: jest.fn() } });
    });
    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado);
    expect(setTouched).toHaveBeenCalledTimes(0);
  });

  test('Debe de llamarse el setValue y no el setTouched, con el espacio final eliminado', async () => {
    // arrange
    const valorEsperado = 'soy una empresa que vende empanadas en cali ';
    const props = {
      name: 'tellUs',
      touched: true,
      value: valorEsperado,
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    await act(async () => {
      wrapper.find('textarea').simulate('blur');
    });
    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(valorEsperado.trimEnd());
    expect(setTouched).toHaveBeenCalledTimes(0);
  });

  test('Debe de llamarse el preventDefault porque el valor ingresado no es valido', () => {
    // arrange
    useFormatter.mockReturnValue([jest.fn(() => ''), changeSelection]);
    const preventDefault = jest.fn();
    const props = {
      name: 'tellUs',
      touched: true,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    wrapper.find('textarea').simulate('beforeInput', { data: '$', preventDefault });
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(1);
  });

  test('Debe de no llamarse el preventDefault porque el valor ingresado es valido', () => {
    // arrange
    const preventDefault = jest.fn();
    const props = {
      name: 'tellUs',
      touched: true,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    wrapper.find('textarea').simulate('beforeInput', { data: 'oli', preventDefault });
    // assert
    expect(preventDefault).toHaveBeenCalledTimes(0);
  });

  test('Debe de mostrarse en rojo debido al error', () => {
    // arrange
    const error = 'Campo obligatorio';
    const props = {
      name: 'tellUs',
      touched: true,
      value: '',
      error,
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    const indicadorError = wrapper.find('.indicador-error').exists();
    const errorMostrado = wrapper.find('span').text().trim();
    // assert
    expect(indicadorError).toBe(true);
    expect(errorMostrado).toBe(error);
  });

  test('Debe de mostrarse como opcional', () => {
    // arrange
    const props = {
      name: 'tellUs',
      touched: true,
      value: '',
      setValue,
      setTouched,
      label: 'Cuéntanos tus dudas...',
      maxlength: 180,
      format: 'textArea',
      optional: true,
    };
    const wrapper = shallow(<TextArea {...props} />);
    // act
    const opcional = wrapper.find('span').text().trim();
    // assert
    expect(opcional).toBe('Opcional');
  });
});
