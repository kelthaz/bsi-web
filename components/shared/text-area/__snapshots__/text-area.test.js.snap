// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente TextArea Debe de mostrarse correctamente el componente 1`] = `
<div
  className="text-area-container"
>
  <div
    className="border-text-area false"
  >
    <textarea
      autoComplete="off"
      className="text-area"
      id="tellUs"
      maxLength={180}
      name="tellUs"
      onBeforeInput={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Cuéntanos tus dudas..."
      tabIndex="0"
      value=""
    />
    <div
      className="resize-text-area"
    />
  </div>
  <span
    className="help-text"
  >
     
  </span>
</div>
`;
